### PR TITLE
jsonbuilder: use serde_json to validate json - v2

### PR DESCRIPTION
**Pull request by jasonish**
Builds on https://github.com/OISF/suricata/pull/10774.

Introduce serde_json as a dev dependency to validate our generated JSON.

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 19942

File: rust/src/smb/smb2_records.rs [hunk 1]
```diff
@@ -695,7 +695,7 @@ mod tests {
         let record: Smb2CreateRequestRecord = result.1;
         assert_eq!(record.disposition, 2); // FILE_CREATE: 2
         assert_eq!(record.create_options, 0x200021);
-        assert_eq!(record.data, &[]);
+        assert_eq!(record.data, &[] as &[u8]);
```
Comment (catenacyber) [inline rust/src/smb/smb2_records.rs:698]: What is happening here ?

File: rust/src/dcerpc/dcerpc.rs [hunk 2]
```diff
@@ -1235,10 +1235,10 @@ pub unsafe extern "C" fn rs_dcerpc_get_tx_cnt(vtx: *mut std::os::raw::c_void) ->
 pub unsafe extern "C" fn rs_dcerpc_get_alstate_progress(tx: *mut std::os::raw::c_void, direction: u8
                                                  )-> std::os::raw::c_int {
     let tx = cast_pointer!(tx, DCERPCTransaction);
-    if direction == Direction::ToServer.into() && tx.req_done {
+    if direction == Direction::ToServer as u8 && tx.req_done {
```
Comment (catenacyber) [inline rust/src/dcerpc/dcerpc.rs:1238]: Why not use `u8::from(` as suggested by the rust people in the linked issue ?

File: rust/src/jsonbuilder.rs [hunk 3]
```diff
@@ -1061,12 +1061,16 @@ mod test {
         assert_eq!(jb.buf, r#"{"first":true"#);
         jb.set_bool("second", false).unwrap();
         assert_eq!(jb.buf, r#"{"first":true,"second":false"#);
+	jb.close().unwrap();
+	let _value: serde_json::Value = serde_json::from_str(&jb.buf).unwrap();
```
Comment (catenacyber) [inline rust/src/jsonbuilder.rs:1065]: Could this be done at a more unique point only if some rust `cfg` is there ?

Side note : we should try to fuzz and eve-validator the eve.json output

Comment (jasonish) [reply rust/src/jsonbuilder.rs:1065]: > Could this be done at a more unique point only if some rust `cfg` is there ?

For what reason?  `serde_json` is already marked as a dev only dependency, so it being linked is already confined to unit tests.

File: rust/src/smb/smb2_records.rs (hunk 1 above)
Comment (jasonish) [reply rust/src/smb/smb2_records.rs:698]: I think Serde introduces some trait impls that confuse the compiler to not knowing the `&[]`.  A better test might might be `assert!(record.data.is_empty())`.

Comment (catenacyber) [reply rust/src/smb/smb2_records.rs:698]: This looks a better test

File: rust/src/jsonbuilder.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/jsonbuilder.rs:1065]: To validate all json output, not only some tests ?

Comment (jasonish) [reply rust/src/jsonbuilder.rs:1065]: Why another `cfg` though? This validation would still just be in test scope right?

Comment (catenacyber) [reply rust/src/jsonbuilder.rs:1065]: No, I meant to run it for every S-V test, not just for unit tests...
(but maybe this is a bad idea)

Comment (jasonish) [reply rust/src/jsonbuilder.rs:1065]: > No, I meant to run it for every S-V test, not just for unit tests... (but maybe this is a bad idea)

S-V test are already subject to JSON validation, so its out of scope here I think.

Comment (catenacyber) [reply rust/src/jsonbuilder.rs:1065]: You said

> Introduce serde_json as a dev dependency to validate our generated JSON.

So, if S-V test are already subject to JSON validation, why do we need more validation ?

Comment (jasonish) [reply rust/src/jsonbuilder.rs:1065]: S-V is not a replacement for unit tests.  We have code that builds JSON which should be unit-tested and validated as well.

**Comment by catenacyber** [comment]
Is there a ticket for [fix serialization of nan and inf](https://github.com/OISF/suricata/pull/10775/commits/71f59e529c3ceaba977ec5d3492bb8b3d3184cd1) ?

**Comment by jasonish** [comment]
> Is there a ticket for [fix serialization of nan and inf](https://github.com/OISF/suricata/pull/10775/commits/71f59e529c3ceaba977ec5d3492bb8b3d3184cd1) ?

From the commit: https://redmine.openinfosecfoundation.org/issues/6921

**Comment by catenacyber** [comment]
> > Is there a ticket for [fix serialization of nan and inf](https://github.com/OISF/suricata/pull/10775/commits/71f59e529c3ceaba977ec5d3492bb8b3d3184cd1) ?
> 
> From the commit: https://redmine.openinfosecfoundation.org/issues/6921

So, already merged...

What are you planning for this draft now ?
I am not sure it is good to add a big dependency just for some testing...

**Comment by jasonish** [comment]
Will close for now, until we have serde_json for some other reason.

**Pull request closed without merging**